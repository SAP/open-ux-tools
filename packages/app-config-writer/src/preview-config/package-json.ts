import { basename, join } from 'node:path';
import { extractYamlConfigFileName } from '../common/package-json';
import { generateVariantsConfig } from '../variants-config';
import type { Editor } from 'mem-fs-editor';
import type { ToolsLogger } from '@sap-ux/logger';
import { type Package, hasDependency } from '@sap-ux/project-access';

/**
 * Ensures that the @sap/ux-ui5-tooling or @sap-ux/preview-middleware dependency exists in the package.json.
 *
 * If none dependency is given, the @sap-ux/preview-middleware will be added as a devDependency.
 *
 * @param fs - file system reference
 * @param basePath - base path to be used for the conversion
 */
export function ensurePreviewMiddlewareDependency(fs: Editor, basePath: string): void {
    const packageJsonPath = join(basePath, 'package.json');
    const packageJson = fs.readJSON(packageJsonPath) as Package | undefined;
    if (!packageJson) {
        return;
    }

    const dependencies = ['@sap-ux/preview-middleware', '@sap/ux-ui5-tooling'];
    if (dependencies.some((dependency) => hasDependency(packageJson, dependency))) {
        return;
    }

    packageJson.devDependencies = { ...packageJson.devDependencies, '@sap-ux/preview-middleware': 'latest' };
    fs.writeJSON(packageJsonPath, packageJson);
}

/**
 * Updates the variants creation script in package.json if needed.
 *
 * If an update is needed, the used intent of the script will be adjusted based on the used UI5 yaml configuration.
 *
 * @param fs - file system reference
 * @param basePath - base path to be used for the conversion
 * @param logger logger to report info to the user
 */
export async function updateVariantsCreationScript(fs: Editor, basePath: string, logger?: ToolsLogger): Promise<void> {
    const packageJsonPath = join(basePath, 'package.json');
    const packageJson = fs.readJSON(packageJsonPath) as Package | undefined;
    if (packageJson?.scripts?.['start-variants-management']) {
        const ui5Yaml = basename(extractYamlConfigFileName(packageJson?.scripts?.['start-variants-management']));
        const yamlPath = join(basePath, ui5Yaml);
        await generateVariantsConfig(basePath, yamlPath, logger, fs);
    }
}
