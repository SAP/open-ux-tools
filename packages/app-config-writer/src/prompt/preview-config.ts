import { prompt, type PromptObject, type Answers } from 'prompts';

/**
 * Prompt if the conversion should be done in simulation.
 *
 * @returns Indicator if the conversion should be simulated.
 */
export async function simulatePrompt(): Promise<boolean> {
    const PROMPT_NAME = 'simulate';
    const question: PromptObject = {
        type: 'confirm',
        name: PROMPT_NAME,
        initial: true,
        message: `The converter renames the local HTML files, deletes the JavaScript and TypeScript files used for the existing preview functionality, and configures virtual endpoints instead.
Do you want to simulate the conversion?`
    };
    const answer = (await prompt([question])) as Answers<typeof PROMPT_NAME>;
    return (
        answer.simulate ?? (await Promise.reject(new Error('An error has occurred. The conversion has been canceled.')))
    ); //in case of doubt, reject
}

/**
 * Prompt if the conversion should include the test runners.
 *
 * @returns Indicator if the conversion should include test runners.
 */
export async function includeTestRunnersPrompt(): Promise<boolean> {
    const PROMPT_NAME = 'includeTests';
    const question: PromptObject = {
        type: 'confirm',
        name: PROMPT_NAME,
        initial: false,
        message: 'Do you want to convert test runners?'
    };
    const answer = (await prompt([question])) as Answers<typeof PROMPT_NAME>;
    return (
        answer.includeTests ??
        (await Promise.reject(new Error('An error has occurred. The conversion has been canceled.')))
    ); //in case of doubt, reject
}
