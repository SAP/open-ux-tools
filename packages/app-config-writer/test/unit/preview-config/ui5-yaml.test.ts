import { create, type Editor } from 'mem-fs-editor';
import { create as createStorage } from 'mem-fs';
import { join } from 'path';
import { ToolsLogger } from '@sap-ux/logger';
import { updatePreviewMiddlewareConfigs, updatePreviewMiddlewareConfig } from '../../../src/preview-config/ui5-yaml';
import type { PreviewConfigOptions } from '../../../src/types';
import type { CustomMiddleware } from '@sap-ux/ui5-config';
import * as projectAccess from '@sap-ux/project-access';

describe('update preview middleware config', () => {
    const logger = new ToolsLogger();
    const warnLogMock = jest.spyOn(ToolsLogger.prototype, 'warn').mockImplementation(() => {});
    const basePath = join(__dirname, '../../fixtures/preview-config');
    let fs: Editor;
    let getAllUi5YamlFileNamesMock: jest.SpyInstance;

    beforeEach(() => {
        jest.clearAllMocks();
        fs = create(createStorage());
        getAllUi5YamlFileNamesMock = jest
            .spyOn(projectAccess, 'getAllUi5YamlFileNames')
            .mockResolvedValue([
                'ui5.yaml',
                'ui5-deprecated-tools-preview.yaml',
                'ui5-deprecated-tools-preview-theme.yaml',
                'ui5-existing-preview-middleware.yaml',
                'ui5-existing-tools-preview.yaml',
                'ui5-no-middleware.yaml',
                'ui5-invalid.yaml'
            ]);
    });

    test('w/o path and intent', async () => {
        const previewMiddleware = {
            name: 'fiori-tools-preview',
            afterMiddleware: 'compression',
            configuration: {
                flp: {
                    path: '/test/flp.html'
                }
            }
        } as CustomMiddleware<PreviewConfigOptions>;
        expect(updatePreviewMiddlewareConfig(previewMiddleware, undefined, undefined)).toMatchSnapshot();
    });

    test('skip yaml configurations not used in any script', async () => {
        const variousConfigsPath = join(basePath, 'various-configs');
        const packageJson = {
            'devDependencies': {
                '@sap-ux/preview-middleware': '0.16.83'
            }
        };
        fs.write(join(variousConfigsPath, 'package.json'), JSON.stringify(packageJson));

        const text = (filename: string) =>
            `The UI5 YAML configuration file '${filename}', is not used in any preview script. Outdated preview middleware will be adjusted, if necessary.`;

        await updatePreviewMiddlewareConfigs(fs, variousConfigsPath, logger);
        expect(fs.read(join(variousConfigsPath, 'package.json'))).toMatchSnapshot();
        expect(warnLogMock).toHaveBeenCalledWith(text('ui5-deprecated-tools-preview.yaml'));
        expect(fs.read(join(variousConfigsPath, 'ui5-deprecated-tools-preview.yaml'))).toMatchSnapshot();
        expect(warnLogMock).toHaveBeenCalledWith(text('ui5-deprecated-tools-preview-theme.yaml'));
        expect(fs.read(join(variousConfigsPath, 'ui5-deprecated-tools-preview-theme.yaml'))).toMatchSnapshot();
        expect(warnLogMock).toHaveBeenCalledWith(text('ui5-existing-preview-middleware.yaml'));
        expect(warnLogMock).toHaveBeenCalledWith(text('ui5-existing-tools-preview.yaml'));
        expect(warnLogMock).toHaveBeenCalledWith(text('ui5-no-middleware.yaml'));
        expect(warnLogMock).toHaveBeenCalledWith(text('ui5.yaml'));
        expect(getAllUi5YamlFileNamesMock).toHaveBeenCalledTimes(1);
    });

    test('skip invalid yaml configurations', async () => {
        const variousConfigsPath = join(basePath, 'various-configs');
        const packageJson = {
            scripts: {
                'invalid':
                    'ui5 serve -o localService/index.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-invalid.yaml'
            },
            'devDependencies': {
                '@sap-ux/preview-middleware': '0.16.83'
            }
        };
        fs.write(join(variousConfigsPath, 'package.json'), JSON.stringify(packageJson));

        await updatePreviewMiddlewareConfigs(fs, variousConfigsPath, logger);
        expect(fs.read(join(variousConfigsPath, 'package.json'))).toMatchSnapshot();
        expect(warnLogMock).toHaveBeenCalledWith(
            `Skipping script 'invalid', which refers to the UI5 YAML configuration file 'ui5-invalid.yaml'. An error occurred when reading 'ui5-invalid.yaml': This file does not comply with the schema.`
        );
    });

    test('skip not found yaml configurations', async () => {
        const variousConfigsPath = join(basePath, 'various-configs');
        const packageJson = {
            scripts: {
                'not:found':
                    'ui5 serve -o localService/index.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-unavailable.yaml'
            },
            'devDependencies': {
                '@sap-ux/preview-middleware': '0.16.83'
            }
        };
        fs.write(join(variousConfigsPath, 'package.json'), JSON.stringify(packageJson));

        await updatePreviewMiddlewareConfigs(fs, variousConfigsPath, logger);
        expect(fs.read(join(variousConfigsPath, 'package.json'))).toMatchSnapshot();
        expect(warnLogMock).toHaveBeenCalledWith(
            `Skipping script 'not:found', because the UI5 YAML configuration file, 'ui5-unavailable.yaml', could not be found.`
        );
    });

    test('no tooling, no middleware', async () => {
        const variousConfigsPath = join(basePath, 'various-configs');
        const packageJson = {
            scripts: {
                'ui:mockserver':
                    'ui5 serve -o /localService/index.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-no-middleware.yaml'
            },
            'devDependencies': {
                '@sap-ux/preview-middleware': '0.16.83'
            }
        };
        fs.write(join(variousConfigsPath, 'package.json'), JSON.stringify(packageJson));

        await updatePreviewMiddlewareConfigs(fs, variousConfigsPath, logger);
        expect(fs.read(join(variousConfigsPath, 'ui5-no-middleware.yaml'))).toMatchSnapshot();
        expect(fs.read(join(variousConfigsPath, 'package.json'))).toMatchSnapshot();
    });

    test('deprecated tools preview with theme', async () => {
        const variousConfigsPath = join(basePath, 'various-configs');
        const packageJson = {
            scripts: {
                'ui:mockserver':
                    'fiori run --open "test/flpSandbox.html?sap-ui-xx-viewCache=false#Chicken-dance" --config ./ui5-deprecated-tools-preview-theme.yaml',
                'start-variants-management': 'ui5 serve --o chicken.html'
            },
            'devDependencies': {
                '@sap/ux-ui5-tooling': '1.15.1'
            }
        };
        fs.write(join(variousConfigsPath, 'package.json'), JSON.stringify(packageJson));

        await updatePreviewMiddlewareConfigs(fs, variousConfigsPath, logger);
        expect(fs.read(join(variousConfigsPath, 'ui5-deprecated-tools-preview-theme.yaml'))).toMatchSnapshot();
        expect(fs.read(join(variousConfigsPath, 'package.json'))).toMatchSnapshot();
    });

    test('deprecated tools preview w/o theme', async () => {
        const variousConfigsPath = join(basePath, 'various-configs');
        const packageJson = {
            scripts: {
                'ui:mockserver':
                    'fiori run --open test/flpSandbox.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-deprecated-tools-preview.yaml',
                'start-variants-management': 'ui5 serve --o chicken.html'
            },
            'devDependencies': {
                '@sap/ux-ui5-tooling': '1.15.1'
            }
        };
        fs.write(join(variousConfigsPath, 'package.json'), JSON.stringify(packageJson));

        await updatePreviewMiddlewareConfigs(fs, variousConfigsPath, logger);
        expect(fs.read(join(variousConfigsPath, 'ui5-deprecated-tools-preview.yaml'))).toMatchSnapshot();
        expect(fs.read(join(variousConfigsPath, 'package.json'))).toMatchSnapshot();
    });

    test('existing tools preview', async () => {
        const variousConfigsPath = join(basePath, 'various-configs');
        const packageJson = {
            scripts: {
                'ui:mockserver':
                    'fiori run -o test/flpSandbox.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-existing-tools-preview.yaml'
            },
            'devDependencies': {
                '@sap/ux-ui5-tooling': '1.15.1'
            }
        };
        fs.write(join(variousConfigsPath, 'package.json'), JSON.stringify(packageJson));

        await updatePreviewMiddlewareConfigs(fs, variousConfigsPath, logger);
        expect(fs.read(join(variousConfigsPath, 'ui5-existing-tools-preview.yaml'))).toMatchSnapshot();
        expect(fs.read(join(variousConfigsPath, 'package.json'))).toMatchSnapshot();
    });

    test('existing preview middleware', async () => {
        const variousConfigsPath = join(basePath, 'various-configs');
        const packageJson = {
            scripts: {
                'ui:mockserver':
                    'fiori run -o localService/index.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-existing-preview-middleware.yaml',
                'ui:opa5':
                    'fiori run -o test/integration/opaTests.qunit.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-existing-preview-middleware.yaml',
                'ui:unit':
                    'fiori run -o test/unit/unitTests.qunit.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-existing-preview-middleware.yaml'
            },
            'devDependencies': {
                '@sap-ux/preview-middleware': '0.16.102'
            }
        };
        fs.write(join(variousConfigsPath, 'package.json'), JSON.stringify(packageJson));

        await updatePreviewMiddlewareConfigs(fs, variousConfigsPath, logger);
        expect(fs.read(join(variousConfigsPath, 'ui5-existing-preview-middleware.yaml'))).toMatchSnapshot();
        expect(fs.read(join(variousConfigsPath, 'package.json'))).toMatchSnapshot();
    });

    test('existing RTA script', async () => {
        const variousConfigsPath = join(basePath, 'various-configs');
        const packageJson = {
            scripts: {
                'start-rta':
                    'ui5 run -o preview.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-existing-preview-middleware.yaml',
                'start-local': 'ui5 run -o /test/flp.html#Chicken-dance --config ./ui5-existing-preview-middleware.yaml'
            },
            'devDependencies': {
                '@sap/ux-ui5-tooling': '1.15.4'
            }
        };
        fs.write(join(variousConfigsPath, 'package.json'), JSON.stringify(packageJson));

        await updatePreviewMiddlewareConfigs(fs, variousConfigsPath, logger);
        expect(fs.read(join(variousConfigsPath, 'ui5-existing-preview-middleware.yaml'))).toMatchSnapshot();
        expect(fs.read(join(variousConfigsPath, 'package.json'))).toMatchSnapshot();
    });

    test('existing start-variants-management and start-control-property-editor script', async () => {
        const variousConfigsPath = join(basePath, 'various-configs');
        const packageJson = {
            scripts: {
                'start-variants-management':
                    'fiori run -o /preview.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-deprecated-tools-preview.yaml',
                'start-control-property-editor':
                    'fiori run -o /editor.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-deprecated-tools-preview.yaml',
                'start-local': 'fiori run -o /test/flp.html#Chicken-dance --config ./ui5-deprecated-tools-preview.yaml'
            },
            'devDependencies': {
                '@sap/ux-ui5-tooling': '1.15.4'
            }
        };
        fs.write(join(variousConfigsPath, 'package.json'), JSON.stringify(packageJson));

        await updatePreviewMiddlewareConfigs(fs, variousConfigsPath, logger);
        expect(fs.read(join(variousConfigsPath, 'ui5-deprecated-tools-preview.yaml'))).toMatchSnapshot();
        expect(fs.read(join(variousConfigsPath, 'package.json'))).toMatchSnapshot();
    });

    test('multiple scripts same yaml configuration', async () => {
        const variousConfigsPath = join(basePath, 'various-configs');
        const packageJson = {
            scripts: {
                'start': 'fiori run -o /test/flp.html#Chicken-dance --config ./ui5-no-middleware.yaml',
                'start-local': 'fiori run -o /test/flp.html#Chicken-dance --config ./ui5-no-middleware.yaml',
                'start-mock': 'fiori run -o /test/flp.html#Chicken-dance --config ./ui5-no-middleware.yaml'
            },
            'devDependencies': {
                '@sap/ux-ui5-tooling': '1.15.4'
            }
        };
        fs.write(join(variousConfigsPath, 'package.json'), JSON.stringify(packageJson));

        await updatePreviewMiddlewareConfigs(fs, variousConfigsPath, logger);

        expect(fs.read(join(variousConfigsPath, 'ui5-no-middleware.yaml'))).toMatchSnapshot();
        expect(fs.read(join(variousConfigsPath, 'package.json'))).toMatchSnapshot();
    });

    test('default ui5.yaml w/o index.html', async () => {
        const variousConfigsPath = join(basePath, 'various-configs');
        const packageJson = {
            scripts: {
                'start': 'fiori run --open "test/flpSandbox.html?sap-ui-xx-viewCache=false#v4lropconvert0711-tile"',
                'start-index': 'fiori run --open "index.html?sap-ui-xx-viewCache=false#v4lropconvert0711-tile"'
            },
            'devDependencies': {
                '@sap/ux-ui5-tooling': '1.15.4'
            }
        };
        fs.write(join(variousConfigsPath, 'package.json'), JSON.stringify(packageJson));

        await updatePreviewMiddlewareConfigs(fs, variousConfigsPath, logger);

        expect(fs.read(join(variousConfigsPath, 'ui5.yaml'))).toMatchSnapshot();
        expect(fs.read(join(variousConfigsPath, 'package.json'))).toMatchSnapshot();
    });

    test('same yaml config same endpoints', async () => {
        fs.write(join(basePath, 'various-configs', 'webapp', 'test', 'flpSandbox.html'), 'dummy content flpSandbox');
        const variousConfigsPath = join(basePath, 'various-configs');
        const packageJson = {
            scripts: {
                'start1': 'fiori run --open "test/flpSandbox.html?sap-ui-xx-viewCache=false#v4lropconvert0711-tile"',
                'start2': 'fiori run --open "test/flpSandbox.html?sap-ui-xx-viewCache=false#v4lropconvert0711-tile"'
            },
            'devDependencies': {
                '@sap/ux-ui5-tooling': '1.15.4'
            }
        };
        fs.write(join(variousConfigsPath, 'package.json'), JSON.stringify(packageJson));

        await updatePreviewMiddlewareConfigs(fs, variousConfigsPath, logger);
        expect(warnLogMock).toHaveBeenCalledTimes(5);
    });

    test('same yaml config different endpoints', async () => {
        const variousConfigsPath = join(basePath, 'various-configs');
        const packageJson = {
            scripts: {
                'start1': 'fiori run --open "test/flpSandbox.html?sap-ui-xx-viewCache=false#v4lropconvert0711-tile"',
                'start2':
                    'fiori run --open "test/flpSandboxMockserver.html?sap-ui-xx-viewCache=false#v4lropconvert0711-tile"'
            },
            'devDependencies': {
                '@sap/ux-ui5-tooling': '1.15.4'
            }
        };
        fs.write(join(variousConfigsPath, 'package.json'), JSON.stringify(packageJson));

        await updatePreviewMiddlewareConfigs(fs, variousConfigsPath, logger);

        expect(fs.read(join(variousConfigsPath, 'ui5.yaml'))).toMatchSnapshot();
        expect(fs.read(join(variousConfigsPath, 'package.json'))).toMatchSnapshot();

        expect(warnLogMock).toHaveBeenCalledWith(
            `Skipping script'start2', because another script also refers to UI5 YAML configuration file, 'ui5.yaml'. Adjust the 'flp.path' property in the UI5 YAML configuration file to the correct endpoint or create a separate UI5 YAML configuration file for script 'start2'. ui5.yaml currently uses test/flpSandbox.html whereas script 'start2' uses 'test/flpSandboxMockserver.html'.`
        );
    });
});
