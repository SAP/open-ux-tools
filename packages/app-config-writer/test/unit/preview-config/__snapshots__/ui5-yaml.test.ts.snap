// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update preview middleware config default ui5.yaml w/o index.html 1`] = `
"# yaml-language-server: $schema=https://sap.github.io/ui5-tooling/schema/ui5.yaml.json

specVersion: \\"3.1\\"
metadata:
  name: v4lropconvert0711
type: application
server:
  customMiddleware:
    - name: fiori-tools-proxy
      afterMiddleware: compression
      configuration:
        ignoreCertError: false # If set to true, certificate errors will be ignored. E.g. self-signed certificates will be accepted
        ui5:
          path:
            - /resources
            - /test-resources
          url: https://ui5.sap.com
        backend:
          - path: /sap
            url: https://sap-ux-mock.com
    - name: fiori-tools-appreload
      afterMiddleware: compression
      configuration:
        port: 35729
        path: webapp
        delay: 300
    - name: fiori-tools-preview
      afterMiddleware: fiori-tools-appreload
      configuration:
        flp:
          theme: sap_horizon
          path: test/flpSandbox.html
          intent:
            object: v4lropconvert0711
            action: tile
"
`;

exports[`update preview middleware config default ui5.yaml w/o index.html 2`] = `"{\\"scripts\\":{\\"start\\":\\"fiori run --open \\\\\\"test/flpSandbox.html?sap-ui-xx-viewCache=false#v4lropconvert0711-tile\\\\\\"\\",\\"start-index\\":\\"fiori run --open \\\\\\"index.html?sap-ui-xx-viewCache=false#v4lropconvert0711-tile\\\\\\"\\"},\\"devDependencies\\":{\\"@sap/ux-ui5-tooling\\":\\"1.15.4\\"}}"`;

exports[`update preview middleware config deprecated tools preview w/o theme 1`] = `
"specVersion: \\"3.1\\"
metadata:
  name: v2lrop0909
type: application
server:
  customMiddleware:
    - name: fiori-tools-preview
      afterMiddleware: compression
      configuration:
        flp:
          path: test/flpSandbox.html
          intent:
            object: Chicken
            action: dance
"
`;

exports[`update preview middleware config deprecated tools preview w/o theme 2`] = `"{\\"scripts\\":{\\"ui:mockserver\\":\\"fiori run --open test/flpSandbox.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-deprecated-tools-preview.yaml\\",\\"start-variants-management\\":\\"ui5 serve --o chicken.html\\"},\\"devDependencies\\":{\\"@sap/ux-ui5-tooling\\":\\"1.15.1\\"}}"`;

exports[`update preview middleware config deprecated tools preview with theme 1`] = `
"specVersion: \\"3.1\\"
metadata:
  name: v2lrop0909
type: application
server:
  customMiddleware:
    - name: fiori-tools-preview
      afterMiddleware: compression
      configuration:
        flp:
          theme: sap_horizon
          path: test/flpSandbox.html
          intent:
            object: Chicken
            action: dance
"
`;

exports[`update preview middleware config deprecated tools preview with theme 2`] = `"{\\"scripts\\":{\\"ui:mockserver\\":\\"fiori run --open \\\\\\"test/flpSandbox.html?sap-ui-xx-viewCache=false#Chicken-dance\\\\\\" --config ./ui5-deprecated-tools-preview-theme.yaml\\",\\"start-variants-management\\":\\"ui5 serve --o chicken.html\\"},\\"devDependencies\\":{\\"@sap/ux-ui5-tooling\\":\\"1.15.1\\"}}"`;

exports[`update preview middleware config existing RTA script 1`] = `
"specVersion: \\"3.1\\"
metadata:
  name: v2lrop0909
type: application
server:
  customMiddleware:
    - name: preview-middleware
      afterMiddleware: compression
      configuration:
        flp:
          path: localService/index.html
          intent:
            object: Chicken
            action: dance
        test:
          - framework: OPA5
            path: /test/integration/opaTests.qunit.html
          - framework: QUnit
            path: /test/unit/unitTests.qunit.html
          - framework: Testsuite
        rta:
          editors:
            - path: preview.html
"
`;

exports[`update preview middleware config existing RTA script 2`] = `"{\\"scripts\\":{\\"start-rta\\":\\"ui5 run -o preview.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-existing-preview-middleware.yaml\\",\\"start-local\\":\\"ui5 run -o /test/flp.html#Chicken-dance --config ./ui5-existing-preview-middleware.yaml\\"},\\"devDependencies\\":{\\"@sap/ux-ui5-tooling\\":\\"1.15.4\\"}}"`;

exports[`update preview middleware config existing preview middleware 1`] = `
"specVersion: \\"3.1\\"
metadata:
  name: v2lrop0909
type: application
server:
  customMiddleware:
    - name: preview-middleware
      afterMiddleware: compression
      configuration:
        flp:
          path: localService/index.html
          intent:
            object: Chicken
            action: dance
        test:
          - framework: OPA5
            path: /test/integration/opaTests.qunit.html
          - framework: QUnit
            path: /test/unit/unitTests.qunit.html
          - framework: Testsuite
        rta:
          editors:
            - path: preview.html
"
`;

exports[`update preview middleware config existing preview middleware 2`] = `"{\\"scripts\\":{\\"ui:mockserver\\":\\"fiori run -o localService/index.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-existing-preview-middleware.yaml\\",\\"ui:opa5\\":\\"fiori run -o test/integration/opaTests.qunit.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-existing-preview-middleware.yaml\\",\\"ui:unit\\":\\"fiori run -o test/unit/unitTests.qunit.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-existing-preview-middleware.yaml\\"},\\"devDependencies\\":{\\"@sap-ux/preview-middleware\\":\\"0.16.102\\"}}"`;

exports[`update preview middleware config existing start-variants-management and start-control-property-editor script 1`] = `
"specVersion: \\"3.1\\"
metadata:
  name: v2lrop0909
type: application
server:
  customMiddleware:
    - name: fiori-tools-preview
      afterMiddleware: compression
      configuration:
        flp:
          intent:
            object: Chicken
            action: dance
"
`;

exports[`update preview middleware config existing start-variants-management and start-control-property-editor script 2`] = `"{\\"scripts\\":{\\"start-variants-management\\":\\"fiori run -o /preview.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-deprecated-tools-preview.yaml\\",\\"start-control-property-editor\\":\\"fiori run -o /editor.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-deprecated-tools-preview.yaml\\",\\"start-local\\":\\"fiori run -o /test/flp.html#Chicken-dance --config ./ui5-deprecated-tools-preview.yaml\\"},\\"devDependencies\\":{\\"@sap/ux-ui5-tooling\\":\\"1.15.4\\"}}"`;

exports[`update preview middleware config existing tools preview 1`] = `
"specVersion: \\"3.1\\"
metadata:
  name: v2lrop0909
type: application
server:
  customMiddleware:
    - name: fiori-tools-preview
      afterMiddleware: compression
      configuration:
        flp:
          theme: sap_horizon
          path: test/flpSandbox.html
          intent:
            object: Chicken
            action: dance
          libs: true
"
`;

exports[`update preview middleware config existing tools preview 2`] = `"{\\"scripts\\":{\\"ui:mockserver\\":\\"fiori run -o test/flpSandbox.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-existing-tools-preview.yaml\\"},\\"devDependencies\\":{\\"@sap/ux-ui5-tooling\\":\\"1.15.1\\"}}"`;

exports[`update preview middleware config multiple scripts same yaml configuration 1`] = `
"specVersion: '4.0'
metadata:
  name: com.sap.cap.fe.ts.sample
  allowSapInternal: true
type: application
framework:
  name: SAPUI5
  version: 1.124.0
server:
  customMiddleware:
    - name: fiori-tools-preview
      afterMiddleware: compression
      configuration:
        flp:
          intent:
            object: Chicken
            action: dance
"
`;

exports[`update preview middleware config multiple scripts same yaml configuration 2`] = `"{\\"scripts\\":{\\"start\\":\\"fiori run -o /test/flp.html#Chicken-dance --config ./ui5-no-middleware.yaml\\",\\"start-local\\":\\"fiori run -o /test/flp.html#Chicken-dance --config ./ui5-no-middleware.yaml\\",\\"start-mock\\":\\"fiori run -o /test/flp.html#Chicken-dance --config ./ui5-no-middleware.yaml\\"},\\"devDependencies\\":{\\"@sap/ux-ui5-tooling\\":\\"1.15.4\\"}}"`;

exports[`update preview middleware config no tooling, no middleware 1`] = `
"specVersion: '4.0'
metadata:
  name: com.sap.cap.fe.ts.sample
  allowSapInternal: true
type: application
framework:
  name: SAPUI5
  version: 1.124.0
server:
  customMiddleware:
    - name: preview-middleware
      afterMiddleware: compression
      configuration:
        flp:
          path: /localService/index.html
          intent:
            object: Chicken
            action: dance
"
`;

exports[`update preview middleware config no tooling, no middleware 2`] = `"{\\"scripts\\":{\\"ui:mockserver\\":\\"ui5 serve -o /localService/index.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-no-middleware.yaml\\"},\\"devDependencies\\":{\\"@sap-ux/preview-middleware\\":\\"0.16.83\\"}}"`;

exports[`update preview middleware config same yaml config different endpoints 1`] = `
"# yaml-language-server: $schema=https://sap.github.io/ui5-tooling/schema/ui5.yaml.json

specVersion: \\"3.1\\"
metadata:
  name: v4lropconvert0711
type: application
server:
  customMiddleware:
    - name: fiori-tools-proxy
      afterMiddleware: compression
      configuration:
        ignoreCertError: false # If set to true, certificate errors will be ignored. E.g. self-signed certificates will be accepted
        ui5:
          path:
            - /resources
            - /test-resources
          url: https://ui5.sap.com
        backend:
          - path: /sap
            url: https://sap-ux-mock.com
    - name: fiori-tools-appreload
      afterMiddleware: compression
      configuration:
        port: 35729
        path: webapp
        delay: 300
    - name: fiori-tools-preview
      afterMiddleware: fiori-tools-appreload
      configuration:
        flp:
          theme: sap_horizon
          path: test/flpSandbox.html
          intent:
            object: v4lropconvert0711
            action: tile
"
`;

exports[`update preview middleware config same yaml config different endpoints 2`] = `"{\\"scripts\\":{\\"start1\\":\\"fiori run --open \\\\\\"test/flpSandbox.html?sap-ui-xx-viewCache=false#v4lropconvert0711-tile\\\\\\"\\",\\"start2\\":\\"fiori run --open \\\\\\"test/flpSandboxMockserver.html?sap-ui-xx-viewCache=false#v4lropconvert0711-tile\\\\\\"\\"},\\"devDependencies\\":{\\"@sap/ux-ui5-tooling\\":\\"1.15.4\\"}}"`;

exports[`update preview middleware config skip invalid yaml configurations 1`] = `"{\\"scripts\\":{\\"invalid\\":\\"ui5 serve -o localService/index.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-invalid.yaml\\"},\\"devDependencies\\":{\\"@sap-ux/preview-middleware\\":\\"0.16.83\\"}}"`;

exports[`update preview middleware config skip multi-file yaml 1`] = `"{\\"scripts\\":{\\"multi-file\\":\\"ui5 serve -o localService/index.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-multi-file.yaml\\"},\\"devDependencies\\":{\\"@sap-ux/preview-middleware\\":\\"0.16.83\\"}}"`;

exports[`update preview middleware config skip not found yaml configurations 1`] = `"{\\"scripts\\":{\\"not:found\\":\\"ui5 serve -o localService/index.html?sap-ui-xx-viewCache=false#Chicken-dance --config ./ui5-unavailable.yaml\\"},\\"devDependencies\\":{\\"@sap-ux/preview-middleware\\":\\"0.16.83\\"}}"`;

exports[`update preview middleware config skip yaml configurations not used in any script 1`] = `"{\\"devDependencies\\":{\\"@sap-ux/preview-middleware\\":\\"0.16.83\\"}}"`;

exports[`update preview middleware config skip yaml configurations not used in any script 2`] = `
"specVersion: \\"3.1\\"
metadata:
  name: v2lrop0909
type: application
server:
  customMiddleware:
    - name: fiori-tools-preview
      afterMiddleware: compression
"
`;

exports[`update preview middleware config skip yaml configurations not used in any script 3`] = `
"specVersion: \\"3.1\\"
metadata:
  name: v2lrop0909
type: application
server:
  customMiddleware:
    - name: fiori-tools-preview
      afterMiddleware: compression
      configuration:
        flp:
          theme: sap_horizon
"
`;

exports[`update preview middleware config w/o path and intent 1`] = `
Object {
  "afterMiddleware": "compression",
  "configuration": Object {
    "flp": Object {
      "path": "/test/flp.html",
    },
  },
  "name": "fiori-tools-preview",
}
`;
